File: projecteuler/102. Triangle containment/main.cpp
#include <bits/stdc++.h>

using namespace std;

// class to give line value for a point and a line
class Line
{
private:
    double x1;
    double y1;
    double x2;
    double y2;
    double slope;

public:
    Line(double x1, double y1, double x2, double y2) : x1(x1), y1(y1), x2(x2), y2(y2)
    {
        if (x1 - x2)
            slope = (y2 - y1) / (x2 - x1);
    }

    bool operator()(double x, double y)
    {
        // if slope exists, use difference in mx - y for two points
        if (x1 - x2)
            return (slope * x - y - slope * x1 + y1) >= 0;
        else
            return (x - x1) >= 0;
    }
};

// solution uses the fact that the centroid of a triangle always lies inside the triangle
// and if origin lies inside the triangle too, both centroid and (0, 0) should lie on the same side of each line
int solve()
{
    ifstream file("p102_triangles.txt");
    int n(0);
    char c;
    double x, y, x1, y1, x2, y2, x3, y3;

    while (!file.eof())
    {
        // read the number and skip the ','
        file >> x1;
        file.get(c);
        file >> y1;
        file.get(c);
        file >> x2;
        file.get(c);
        file >> y2;
        file.get(c);
        file >> x3;
        file.get(c);
        file >> y3;
        file.get(c);

        // get the centroid
        x = (x1 + x2 + x3) / 3;
        y = (y1 + y2 + y3) / 3;

        // construct three line objects
        Line line1(x1, y1, x2, y2), line2(x2, y2, x3, y3), line3(x3, y3, x1, y1);

        // check if centroid and origin lie in the same side of the lines for all lines
        if (line1(x, y) == line1(0, 0) && line2(x, y) == line2(0, 0) && line3(x, y) == line3(0, 0))
            n++;
    }

    return n;
}

int main()
{
    cout << solve(); // answer: 228
}

File: projecteuler/102. Triangle containment/main.py
'''
The solution uses the fact that the centroid of a triangle always lies inside the triangle
and if origin lies inside the triangle too, both centroid and (0, 0) should lie on the same side of each line
'''


# function to give line value for a point and a line
def line_value(x1, y1, x2, y2):
    # if slope exists, use difference in mx - y for two points
    if x1 - x2:
        slope = (y2 - y1) / (x2 - x1)
        return lambda x, y: (slope * x - y - slope * x1 + y1) >= 0
    else:
        return lambda x, y: (x - x1) >= 0


def solver():
    n = 0
    with open('p102_triangles.txt') as file:
        lines = file.readlines()

    for line in lines:
        # read a single line
        line_point = line.split(",")
        x1, y1, x2, y2, x3, y3 = [int(x) for x in line_point]

        # calculate the centroid
        x, y = (x1 + x2 + x3) / 3, (y1 + y2 + y3) / 3

        # use line values for all three lines that form the triangle
        line1 = line_value(x1, y1, x2, y2)
        line2 = line_value(x2, y2, x3, y3)
        line3 = line_value(x3, y3, x1, y1)

        # check if centroid and origin lie in the same side of the lines for all lines
        if line1(x, y) == line1(0, 0) and line2(x, y) == line2(0, 0) and line3(x, y) == line3(0, 0):
            n += 1
    return n


print(solver()) # answer: 228
